pkg/repo/fs.go

```diff
@@ -0,0 +1,324 @@
+			}
+			if err = pem.Encode(file, &keyBlock); err != nil {
+				return "", fmt.Errorf("failed to encode key file: %w", err)
+			}
+			if err = file.Close(); err != nil {
+				return "", fmt.Errorf("failed to close key file: %w", err)
+			}
+			if err = os.Chmod(keyFile, util.OS_USER_RW); err != nil {
+				return "", fmt.Errorf("failed to set permission on key file: %w", err)
+			}
+		} else {
+			return "", fmt.Errorf("failed to stat user ID key '%s': %w",
+				keyFile, err)
+		}
+	}
+
+	return keyFile, nil
+}
+
+func ensureLibp2pKey(configDir string) (string, error) {
+	keyName := "libp2p_private_key"
+
+	// We include the port in the filename so that in devstack multiple nodes
+	// running on the same host get different identities
+	privKeyPath := filepath.Join(configDir, keyName)
+
+	if _, err := os.Stat(privKeyPath); errors.Is(err, os.ErrNotExist) {
+		// Private key does not exist - create and write it
+
+		// Creates a new RSA key pair for this host.
+		prvKey, _, err := crypto.GenerateKeyPairWithReader(crypto.RSA, bitsPerKey, rand.Reader)
+		if err != nil {
+			log.Error().Err(err)
+			return "", err
+		}
+
+		keyOut, err := os.OpenFile(privKeyPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, util.OS_USER_RW)
+		if err != nil {
+			return "", fmt.Errorf("failed to open key.pem for writing: %v", err)
+		}
+		privBytes, err := crypto.MarshalPrivateKey(prvKey)
+		if err != nil {
+			return "", fmt.Errorf("unable to marshal private key: %v", err)
+		}
+		// base64 encode privBytes
+		b64 := base64.StdEncoding.EncodeToString(privBytes)
+		_, err = keyOut.WriteString(b64 + "\n")
+		if err != nil {
+			return "", fmt.Errorf("failed to write to key file: %v", err)
+		}
+		if err := keyOut.Close(); err != nil {
+			return "", fmt.Errorf("error closing key file: %v", err)
+		}
+		log.Debug().Msgf("wrote %s", privKeyPath)
+	} else {
+		return "", err
+	}
+
+	// Now that we've ensured the private key is written to disk, read it! This
+	// ensures that loading it works even in the case where we've just created
+	// it.
+
+	// read the private key
+	keyBytes, err := os.ReadFile(privKeyPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to read private key: %v", err)
+	}
+	// base64 decode keyBytes
+	b64, err := base64.StdEncoding.DecodeString(string(keyBytes))
+	if err != nil {
+		return "", fmt.Errorf("failed to decode private key: %v", err)
+	}
+	// parse the private key
+	_, err = crypto.UnmarshalPrivateKey(b64)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse private key: %v", err)
+	}
+
+	return privKeyPath, nil
+}
+
+const defaultRunInfoFilename = "bacalhau.run"
+const runInfoFilePermissions = 0755
+
+func (fsr *FsRepo) WriteRunInfo(ctx context.Context, summaryShellVariablesString string) (string, error) {
+	runInfoPath := filepath.Join(fsr.path, defaultRunInfoFilename)
+
+	// TODO kill this
+	devStackRunInfoPath := os.Getenv("DEVSTACK_ENV_FILE")
+	if devStackRunInfoPath != "" {
+		runInfoPath = devStackRunInfoPath
+	}
```

Comment:
I think we should pick a reasonable port number and default to it, it would also be a nice-to-have if the UserKey, Libp2p key, and IPFS node key were deterministically generated.
_________________________
However, I am torn on what the devstack should be. Part of me likes the simplicity of spinning up all the nodes as a single process - it's a good utility for simple test cases. On the other hand I find it a little confusing - creating N number of nodes some of which are compute, or requesters, or both (hybrid) that share a repo and associate config seems wrong. At present, we are treating the repo and config as a global value accessed (somewhat) sporadically through out the code (see the number of places we call `GetAPIClient()`, or access `system.Env`), and I have low confidence things will continue to work as expected as we continue ripping off the plaster. What happens today if I make a devstack with 3 requester nodes? Which API am I talking to? 

Treating the config, repo, and environment (`pkg/system/environment`) as global values doesn't lend itself well to the in-memory nature of the devstack. And as you've suggested [here](https://github.com/bacalhau-project/bacalhau/pull/2740#discussion_r1296985812) we should move away from depending directly on the config package, and instead pass the configuration values in at run-time. I think that shift will help with things, but I am also thinking we want a different thingy - a devcluster? That lets us to do things like:
1. Compute node with a docker executor
2. Compute node with a wasm executor
3. Compute node with a pluggable python executor
4. Configure a compute node with some unique storage implementation
5. Configure a compute with some bespoke job selection policy
6. Compute node with networking
7. Create a Requester connected to some compute nodes
8. Create a Requester connected to others
9. Create a Requester+Compute node
10.  ... other node configuration stuff... get where I am going with this?

Spin up all of these nodes as separate processes (`os.Exec` or better, docker (although a docker executor in a docker container would be..hard?), fire a job off to one of them and make assertions on expected behavior, or try out a new feature being dev'd on.

Reply:
I also suffer the same cognitive overload when running hybrid clusters, particularly around the node setup (where half of it is shared and then we diverge based on a boolean flag).  I had thought at one point I'd like to `docker compose` a mini cluster where the docker socket was mounted inside the compute node containers (so it could still run docker images outside docker) but I think that's a bit clunkier than I'd prefer and is fiddly to manage.  I guess there's nothing actually stopping me finding a simple process supervisor and configuring a mini cluster for dev, with specific ports etc (assuming bootstrap is easy enough) but it feels like a lost opportunity with devstack.  

This is probably best addressed in another PR though at this point.

Reply:
> devstack, with all the limitations mentioned above, is one of the best things we have in the project!

I could not agree more! I've used it a lot here as well. However [the](https://github.com/bacalhau-project/bacalhau/blob/main/pkg/devstack/devstack.go#L96) [cognitive](https://github.com/bacalhau-project/bacalhau/blob/main/pkg/devstack/devstack.go#L104) [load](https://github.com/bacalhau-project/bacalhau/blob/main/pkg/devstack/devstack.go#L129) [to](https://github.com/bacalhau-project/bacalhau/blob/main/pkg/devstack/devstack.go#L175-L176) [understand](https://github.com/bacalhau-project/bacalhau/blob/main/pkg/devstack/devstack.go#L212) [how](https://github.com/bacalhau-project/bacalhau/blob/main/pkg/devstack/devstack.go#L164) [it](https://github.com/bacalhau-project/bacalhau/blob/main/pkg/devstack/devstack.go#L215) [sets](https://github.com/bacalhau-project/bacalhau/blob/main/pkg/devstack/devstack.go#L357-L359) up its nodes is quite high and ripe for a refactor at some point. I am not saying "devstack is bad", just pointing out ways of improving it.

> I don't understand why there will be a confusion when running with multiple requester nodes

Maybe confusion wasn't the right word here, but an area that frustrates me centers around keys and signatures, and we can 100% fix this with more dep injection: 

Assume we have a devstack with 3 requester nodes. True, they all have different API ports (although we only [advertise this via a debug log](https://github.com/bacalhau-project/bacalhau/blob/main/pkg/devstack/devstack.go#L339C3-L339C12), which is off by default), but they all have the [same identity](https://github.com/bacalhau-project/bacalhau/blob/main/pkg/requester/publicapi/client.go#L265) due to the [global nature](https://github.com/bacalhau-project/bacalhau/blob/main/pkg/system/config.go#L165) of the ClientID. This wouldn't be the case if either 1) they were different processes or 2) we passed the clientID to `NewNode`.